extension-androidauto/src/main/java/com/mapbox/maps/extension/androidauto/MapboxCarMapObserver.kt

```diff
@@ -0,0 +1,120 @@
+package com.mapbox.maps.extension.androidauto
+
+import android.graphics.Rect
+import androidx.car.app.SurfaceCallback
+import com.mapbox.maps.EdgeInsets
+import com.mapbox.maps.MapboxExperimental
+import com.mapbox.maps.ScreenCoordinate
+
+/**
+ * Many downstream services will not work until the surface has been created and the map has
+ * loaded. This interface allows you to create custom Mapbox experiences for the car.
+ */
+@MapboxExperimental
+interface MapboxCarMapObserver {
+
+  /**
+   * Called when a [MapboxCarMapSurface] has been loaded.
+   * You can assume there will only be a single surface at a time.
+   *
+   * @see [MapboxCarMap.registerObserver]
+   *
+   * @param mapboxCarMapSurface loaded and ready to use car map surface
+   */
+  fun onAttached(mapboxCarMapSurface: MapboxCarMapSurface) {
+    // No op by default
+  }
+
+  /**
+   * Called when a [MapboxCarMapSurface] has been detached from this observer. Some examples that
+   * can cause this to detach are:
+   * - [MapboxCarMap] lifecycle is destroyed
+   * - This observer has been unregistered with [MapboxCarMap.unregisterObserver]
+   *
+   * You can assume that there was a corresponding call to [onAttached] with the same
+   * [MapboxCarMapObserver] instance.
+   *
+   * @param mapboxCarMapSurface loaded and ready car map surface
+   */
+  fun onDetached(mapboxCarMapSurface: MapboxCarMapSurface) {
+    // No op by default
+  }
+
+  /**
+   * Called when the car library updates the visible regions for the surface. For example, this
+   * is triggered when the action buttons come in and out of visibility.
+   * You can assume this will be called after [onAttached].
+   *
+   * @param visibleArea the visible area provided by the host
+   * @param edgeInsets distance from each side of the screen that creates the [visibleArea]
+   */
+  fun onVisibleAreaChanged(visibleArea: Rect, edgeInsets: EdgeInsets) {
+    // No op by default
+  }
+
+  /**
+   * Allows you to implement or observe the map scroll gesture handler. The surface is [onAttached]
+   * before this can be triggered.
+   *
+   * @see [SurfaceCallback.onScroll] for instructions to enable.
+   *
+   * @param mapboxCarMapSurface loaded and ready car map surface
+   * @param distanceX the distance in pixels along the X axis
+   * @param distanceY the distance in pixels along the Y axis
+   *
+   * @return true when the fling scroll was handled, false will trigger the default handler
+   */
+  fun onScroll(
+    mapboxCarMapSurface: MapboxCarMapSurface,
+    distanceX: Float,
+    distanceY: Float
+  ): Boolean {
+    // By default, scroll is handled internally
+    return false
+  }
```

Comment:
would user be able to provide their own logic? so this on Scroll will actually overwrite the internal behaviour? And if multiple MapboxCarMapObserver is registered, what's the sequence of the overwrite? will it be the first ever onScroll that returns true?

Reply:
Yes, downstream can specify their own logic. If `any` of the observers return `true`, the scroll is considered handled.

``` kotlin
internal class CarMapSurfaceOwner {
..
val handled = carMapObservers.any { it.scroll(carMapSurface, distanceX, distanceY) }
if (handled) return
```

Considered `first`, but went with `any` to ensure all observers are called. If a developer has multiple observers that handle the event, that is a behavior they are creating.

Reply:
do we have some way to workaround it? otherwise I think we should mention it somehow in the documentation, as it would easily introduce bugs if not handled properly.

The similar pattern I think is the OnClick handling on Android, but in that case, there's only one observer(inherited though) so that any layer can return true and overwrite the default behaviour, and the order is defined as well, since it's based on inheritance.

Reply:
One other idea, shall we make it an abstract class, where users can provide overwrites? and make it settable only once?

Reply:
Yeah setting it once is similar to how `setOnTouchListener` works. And I did consider this approach but it was more complicated and there are trade-offs. https://developer.android.com/training/gestures/detector

We want to provide a default implementation. We also want to provide a way to override one or all of the gesture. So maybe we add 3 gesture detectors with their own default implementation? 

I have thought about this, and it's the `onFling` that becomes complicated. Because it onScale should cancel the onFling gesture. 

An default and open class could work (not an abstract class because it will be set by default). That could be an alternative. 🤔

Reply:
I started making a `open class MapboxCarMapGestures`. That seems like a good approach

it will take a bit to shuffle the tests around. So we can either merge and have another review, or wait to have the refactor finish. What do you think @pengdev ?

Reply:
ok, also good to have document the explanation https://github.com/mapbox/mapbox-maps-android/issues/1152

resolving this thread. I'll continue on my branch to make it so the iteration is ready soon. Resolving this thread